geniza/corpus/models.py

```diff
@@ -1337,38 +1337,50 @@ def merge_with(self, merge_docs, rationale, user=None):
+            # check for match. for each pair of footnotes, there are two possible cases for
+            # non-equivalence:
+            # - the footnote to be merged in has annotations
+            # - there are fields on the footnotes that don't match
+            # in the former case, merge the two by migrating the annotations from the footnote to
+            # be merged in to an otherwise matching footnote if there is one; else, add it to doc.
+            # in the latter case, simply add the footnote to this document.
+
             if footnote.annotation_set.exists():
-                # if the footnote has annotations, reassign them to matching/created footnote
-                (self_fn, _) = self.footnotes.get_or_create(
-                    # for multiselect field list, need to cast to list to compare
-                    doc_relation__in=list(footnote.doc_relation),
-                    source_id=footnote.source.pk,
-                    # TODO: What to do if found, but location or notes don't match? Discard?
-                    defaults={
-                        "notes": footnote.notes,
-                        "location": footnote.location,
-                        "url": footnote.url,
-                    },
-                )
-                # reassign each annotation's footnote to the footnote on the merge result doc
-                for annotation in footnote.annotation_set.all():
-                    annotation.footnote = self_fn
-                    annotation.save()
-            elif not equiv_fn:
-                # if there is otherwise not a match, add the footnote to the merge result document
+                try:
+                    # if the footnote to be merged in has annotations, try to reassign them to an
+                    # otherwise matching footnote to avoid unique constraint violation
+                    self_fn = self.footnotes.get(
+                        # for multiselect field list, need to cast to list to compare
+                        doc_relation__in=list(footnote.doc_relation),
+                        source_id=footnote.source.pk,
+                    )
+                    # copy over notes, location, url if missing from self_fn
+                    for attr in ["notes", "location", "url"]:
+                        if not getattr(self_fn, attr) and getattr(footnote, attr):
+                            setattr(self_fn, attr, getattr(footnote, attr))
+                    self_fn.save()
+                    # reassign each annotation's footnote to the footnote on this doc
+                    for annotation in footnote.annotation_set.all():
+                        annotation.footnote = self_fn
+                        annotation.save()
+                except Footnote.DoesNotExist:
+                    # if there is no match, we are clear of any unique constaint violation and can
+                    # simply add the footnote to this document
+                    self.footnotes.add(footnote)
```

Comment:
I think pulling this out separately (instead of using get or create) is clearer & better